src/urh/dev/PCAPNG.py

```diff
@@ -0,0 +1,114 @@
+        shb = struct.pack(">IIIHHQ",
+            BLOCKTYPE,
+            total_block_len,
+            MAGIC_NUMBER,
+            VERSION_MAJOR, VERSION_MINOR,
+            SECTIONLENGTH)
+
+        if (shb_userappl != ""):
+            SHB_USERAPPL = 4
+            strpad = shb_userappl.ljust(shb_userappl_padded_len, "\0")
+            shb += struct.pack(">HH", SHB_USERAPPL, shb_userappl_padded_len)
+            shb += bytes(strpad, 'ascii')
+
+        if (shb_hardware != ""):
+            SHB_HARDWARE = 2
+            strpad = shb_hardware.ljust(shb_hardware_padded_len, "\0")
+            shb += struct.pack(">HH", SHB_HARDWARE, shb_hardware_padded_len)
+            shb += bytes(strpad, 'ascii')
+        
+        shb += struct.pack(">I",total_block_len)
+        return shb
+
+    def __build_pcapng_idb(self, link_type) -> bytes:
+        BLOCKTYPE = 0x00000001
+        BLOCKLENGTH = 20
+        SNAP_LEN = 0
+
+        return struct.pack(">IIHHII", 
+            BLOCKTYPE,
+            BLOCKLENGTH,
+            link_type, 0,
+            SNAP_LEN,
+            BLOCKLENGTH)
+
+    def __build_pcapng_epb(self, packet : bytes, timestamp : float) -> bytes:
+        BLOCKTYPE = 0x00000006
+        BLOCKHEADERLEN = 32
+        INTERFACE_ID = 0
+
+        captured_packet_len = len(packet)
+        original_packet_len = captured_packet_len
+        padded_packet_len = math.ceil(captured_packet_len/4) * 4
+        padding_len = padded_packet_len - original_packet_len
+        padded_packet = packet + bytearray(padding_len)
+        block_total_length = BLOCKHEADERLEN + padded_packet_len
+        timestamp_int = int(timestamp *1e6) # Set the proper resolution
+        timestamp_high = timestamp_int >> 32
+        timestamp_low = timestamp_int & 0x00000000FFFFFFFF
+
+        epb = struct.pack(">IIIIIII", 
+            BLOCKTYPE,
+            block_total_length,
+            INTERFACE_ID,
+            timestamp_high,
+            timestamp_low,
+            captured_packet_len,
+            original_packet_len)
+        epb += padded_packet
+        epb += struct.pack(">I", block_total_length)
+        return epb
+
+    ############################################################################################################
+    # PUBLIC INTERFACES
+    def create_pcapng_file(self, filename : str, shb_userappl : str = "", shb_hardware : str = "", link_type : int =147) -> bytes:
+        if (filename == ""):
+            return
+        self.__filename = filename
+
+        shb_bytes = self.__build_pcapng_shb(shb_userappl,shb_hardware)
+        idb_bytes = self.__build_pcapng_idb(link_type)
+
+        if os.path.isfile(self.__filename):
+            logger.warning("{0} already exists. Overwriting it".format(filename))
+
+        with open(self.__filename, "wb") as f:
+            f.write(shb_bytes)
+            f.write(idb_bytes)
+
+    def pcapng_file_append_packet(self, packet : bytes, timestamp: float):
+        with open(self.__filename, "ab") as f:
```

Comment:
This requires opening and closing the PCAP file for each packet. It would be nice to have a method performing this in a batch.

Reply:
Missed this comment. I just fixed it. I added **pcapng_file_append_multi_packets** function that optimizes this. I wanted to leave PCAPNG class isolated from Messages or any other specific class, so I'm passing a generic list of [packet,timestamp]. Let me know if it's ok for you.